### DISCUSSION TITLE
Expectation value with marginal.

### DISCUSSION THREAD

**Discussion by yapolyak** [NVIDIA/cuQuantum#116] [Q&A]
Hi,

It seems that while there is a high-level API to calculate expectation value from a `tensorNetworkState` and a `tensorNetworkOperator` (https://docs.nvidia.com/cuda/cuquantum/latest/cutensornet/api/functions.html#cutensornetcreateexpectation) but not from a `tensorNetworkMarginal` - is that right? I just was wondering if (re-)using a reduced density matrix to calculate expectation values might be more memory-efficient than using state objects, although it depends on an operator I guess (as in how many modes you'd need). I understand that one could do the necessary contractions manually, but it could be nice to have a dedicated API call for that.
Thanks!

**Comment by yangcal** [comment]
Hello, 

Can you provide more details on the problem specification? Is the goal to compute expectation values for different `tensorNetworkOperator` (indivisually computed or summed?) using the same `tensorNetworkMarginal` object (the marginal modes fixed)? 

If only one expectation value is requested, directly using `Expectation` APIs should be faster than `Marginal` + manual contraction as `Expectation` allows for more space for path optimization. 

@DmitryLyakh for visibility too

**Reply by DmitryLyakh** [reply]
In practice, storing reduced density matrices (marginals) and reusing them could indeed make sense for 1-local and 2-local operators with diverse (X,Y,Z) terms. Can you describe your use case where such a feature would be important for performance?

**Reply by yapolyak** [reply]
Hi @yangcal , hi @DmitryLyakh ,

Thank you for your replies and apologies for a pause taken!

To be honest I was thinking about a general iterative (VQE) procedure, where only limited set of gate parameters changes through the iterations, and the same Hamiltonian is evaluated every iteration to get the energy. Of course, as I think you suggest, 4-local chemical Hamiltonian might be too large so that one would need as many pre-contracted (or on the order) density matrices as number of operator commuting sets... But I still was thinking of a general API where one could precompute a set of them, and then use them in an iterative procedure to obtain energy at every step.

As far as I understand, `ExpectationValue` is contracted in full every time it is told to be computed right? So what is the point of the mutable gates then in the `State` object?

I guess I am mainly thinking loudly at this point (although I am writing a wrapper for your high-level API atm for our extension).

Many thanks!
